### ISSUE TITLE
Bogus assertions in ContainerParentDataMixin.detach()

### ISSUE THREAD

**Issue by kgdev** [flutter/flutter#22105]
`ContainerParentDataMixin  assert(previousSibling != this);`

previousSibling is an instance of ChildType, while "this" is an intance of ParentData.

So it can never be true?

**Comment by zoechi** [comment]
Please add the output of `flutter doctor -v` and the code that allows to reproduce
and the exact and full error output.

**Comment by zoechi** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. We are therefore reluctantly going to close this bug for now. Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!
Thanks for your contribution.

**Comment by Hixie** [comment]
@kgdev is right, this code is completely bogus.

@goderbauer I'm not sure how this code can ever even be running, because the very next assert is `previousSiblingParentData.nextSibling == this` which makes even less sense: how can the `nextSibling` (which is a `RenderObject`) ever equal `this`, which is a `ParentData`??

I don't see any reports of this assert being hit, so presumably this code is never being run. Maybe by the time we detach `parentData`s, all the children have been detached first? Assuming that is indeed happening, we should decide whether to just remove this code and replace it with just `assert(previousSibling == null)`, requiring it to happen.

**Comment by Hixie** [comment]
Another reason to just remove this code is that it doesn't update the parent's child count.

**Comment by goderbauer** [comment]
Looks like most (all?) `ContainerParentDataMixin`s get their sibling pointers cleared here (and the parent's count is updated correctly as well there):

https://github.com/flutter/flutter/blob/13382f41c06acca01bcfd24725ef7d6ef9d368cf/packages/flutter/lib/src/rendering/object.dart#L2990-L3012

This is called from `ContainerRenderObjectMixin.remove(ChildType child)` before calling `RenderObject.dropChild(ChildType child)`. The latter then calls `ContainerParentDataMixin.detach()`, which has those questionable asserts in it. At that point the siblings pointers are already null and those asserts never fire.

I agree, we should just remove them and require instead in that detach that the pointers are already nulled out.
